File: src-tauri/src/s2_lite/canonical.rs
use super::types::{CommitDot, CommitRef, EntityKey};
use chrono::NaiveDate;
use serde::Serialize;
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::cmp::Ordering;
use uuid::{Variant, Version};

const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;

pub const I64_MIN: i128 = i64::MIN as i128;
pub const I64_MAX: i128 = i64::MAX as i128;
pub const U64_MAX: u128 = u64::MAX as u128;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ProtocolError(pub &'static str);

pub type Result<T> = std::result::Result<T, ProtocolError>;

pub fn jcs_bytes<T: Serialize>(value: &T) -> Result<Vec<u8>> {
    let value =
        serde_value::to_value(value).map_err(|_| ProtocolError("jcs_serialization_failed"))?;
    let mut output = String::new();
    write_jcs_value(&value, &mut output)?;
    Ok(output.into_bytes())
}

fn write_jcs_string(value: &str, output: &mut String) -> Result<()> {
    let encoded =
        serde_json::to_string(value).map_err(|_| ProtocolError("jcs_serialization_failed"))?;
    output.push_str(&encoded);
    Ok(())
}

fn write_jcs_float(value: f64, output: &mut String) -> Result<()> {
    if !value.is_finite() {
        return Err(ProtocolError("jcs_non_finite_number"));
    }
    if value == 0.0 {
        output.push('0');
    } else {
        output.push_str(ryu_js::Buffer::new().format_finite(value));
    }
    Ok(())
}

fn write_jcs_unsigned(value: u64, output: &mut String) -> Result<()> {
    if value > MAX_SAFE_INTEGER {
        return Err(ProtocolError("jcs_unsafe_integer"));
    }
    output.push_str(&value.to_string());
    Ok(())
}

fn write_jcs_signed(value: i64, output: &mut String) -> Result<()> {
    if value.unsigned_abs() > MAX_SAFE_INTEGER {
        return Err(ProtocolError("jcs_unsafe_integer"));
    }
    output.push_str(&value.to_string());
    Ok(())
}

fn compare_utf16(a: &str, b: &str) -> Ordering {
    a.encode_utf16().cmp(b.encode_utf16())
}

fn write_jcs_value(value: &serde_value::Value, output: &mut String) -> Result<()> {
    use serde_value::Value as SValue;
    match value {
        SValue::Bool(value) => output.push_str(if *value { "true" } else { "false" }),
        SValue::U8(value) => write_jcs_unsigned((*value).into(), output)?,
        SValue::U16(value) => write_jcs_unsigned((*value).into(), output)?,
        SValue::U32(value) => write_jcs_unsigned((*value).into(), output)?,
        SValue::U64(value) => write_jcs_unsigned(*value, output)?,
        SValue::I8(value) => write_jcs_signed((*value).into(), output)?,
        SValue::I16(value) => write_jcs_signed((*value).into(), output)?,
        SValue::I32(value) => write_jcs_signed((*value).into(), output)?,
        SValue::I64(value) => write_jcs_signed(*value, output)?,
        SValue::F32(value) => write_jcs_float(f64::from(*value), output)?,
        SValue::F64(value) => write_jcs_float(*value, output)?,
        SValue::Char(value) => write_jcs_string(&value.to_string(), output)?,
        SValue::String(value) => write_jcs_string(value, output)?,
        SValue::Unit | SValue::Option(None) => output.push_str("null"),
        SValue::Option(Some(value)) | SValue::Newtype(value) => write_jcs_value(value, output)?,
        SValue::Seq(values) => {
            output.push('[');
            for (index, value) in values.iter().enumerate() {
                if index != 0 {
                    output.push(',');
                }
                write_jcs_value(value, output)?;
            }
            output.push(']');
        }
        SValue::Map(values) => {
            let mut entries = values
                .iter()
                .map(|(key, value)| match key {
                    SValue::String(key) => Ok((key.as_str(), value)),
                    _ => Err(ProtocolError("jcs_non_string_object_key")),
                })
                .collect::<Result<Vec<_>>>()?;
            entries.sort_by(|(left, _), (right, _)| compare_utf16(left, right));
            output.push('{');
            for (index, (key, value)) in entries.iter().enumerate() {
                if index != 0 {
                    output.push(',');
                }
                write_jcs_string(key, output)?;
                output.push(':');
                write_jcs_value(value, output)?;
            }
            output.push('}');
        }
        SValue::Bytes(_) => return Err(ProtocolError("jcs_unsupported_value")),
    }
    Ok(())
}

pub fn sha256_hex(bytes: &[u8]) -> String {
    format!("{:x}", Sha256::digest(bytes))
}

pub fn sha256_jcs<T: Serialize>(value: &T) -> Result<String> {
    Ok(sha256_hex(&jcs_bytes(value)?))
}

pub fn validate_canonical_uuid(value: &str) -> Result<()> {
    let parsed =
        uuid::Uuid::parse_str(value).map_err(|_| ProtocolError("invalid_canonical_uuid"))?;
    if parsed.to_string() != value
        || parsed.get_variant() != Variant::RFC4122
        || parsed.get_version().is_none()
    {
        return Err(ProtocolError("invalid_canonical_uuid"));
    }
    Ok(())
}

pub fn validate_canonical_uuid_v4(value: &str) -> Result<()> {
    validate_canonical_uuid(value)?;
    let parsed =
        uuid::Uuid::parse_str(value).map_err(|_| ProtocolError("invalid_canonical_uuid_v4"))?;
    if parsed.get_version() != Some(Version::Random) {
        return Err(ProtocolError("invalid_canonical_uuid_v4"));
    }
    Ok(())
}

fn canonical_unsigned_decimal(value: &str) -> bool {
    value == "0"
        || (!value.is_empty()
            && !value.starts_with('0')
            && value.as_bytes().iter().all(u8::is_ascii_digit))
}

pub fn parse_writer_seq(value: &str) -> Result<u64> {
    if !canonical_unsigned_decimal(value) {
        return Err(ProtocolError("invalid_writer_seq"));
    }
    let parsed = value
        .parse::<u64>()
        .map_err(|_| ProtocolError("invalid_writer_seq"))?;
    if parsed == 0 {
        return Err(ProtocolError("invalid_writer_seq"));
    }
    Ok(parsed)
}

pub fn parse_int64_decimal(value: &str, minimum: i64, maximum: i64) -> Result<i64> {
    if value == "-0"
        || value.starts_with('+')
        || value.is_empty()
        || (value.starts_with('0') && value.len() > 1)
        || (value.starts_with("-0") && value.len() > 2)
        || !value
            .strip_prefix('-')
            .unwrap_or(value)
            .as_bytes()
            .iter()
            .all(u8::is_ascii_digit)
    {
        return Err(ProtocolError("invalid_int64_decimal_string"));
    }
    let parsed = value
        .parse::<i64>()
        .map_err(|_| ProtocolError("int64_out_of_range"))?;
    if parsed < minimum || parsed > maximum {
        return Err(ProtocolError("int64_out_of_range"));
    }
    Ok(parsed)
}

pub fn validate_safe_integer(value: &Value, minimum: i64, maximum: i64) -> Result<i64> {
    let parsed = value
        .as_f64()
        .ok_or(ProtocolError("invalid_safe_integer"))?;
    if !parsed.is_finite()
        || parsed.fract() != 0.0
        || parsed.abs() > MAX_SAFE_INTEGER as f64
        || parsed < minimum as f64
        || parsed > maximum as f64
    {
        return Err(ProtocolError("invalid_safe_integer"));
    }
    Ok(parsed as i64)
}

pub fn validate_float64(value: &Value, minimum: f64, maximum: f64) -> Result<f64> {
    let parsed = value.as_f64().ok_or(ProtocolError("invalid_float64"))?;
    if !parsed.is_finite() || parsed < minimum || parsed > maximum {
        return Err(ProtocolError("invalid_float64"));
    }
    Ok(parsed)
}

pub fn validate_date(value: &str) -> Result<()> {
    let bytes = value.as_bytes();
    if bytes.len() != 10
        || bytes[4] != b'-'
        || bytes[7] != b'-'
        || ![0..4, 5..7, 8..10]
            .into_iter()
            .all(|range| bytes[range].iter().all(u8::is_ascii_digit))
    {
        return Err(ProtocolError("invalid_date"));
    }
    let year = value[0..4].parse::<i32>().expect("ASCII digits");
    let month = value[5..7].parse::<u32>().expect("ASCII digits");
    let day = value[8..10].parse::<u32>().expect("ASCII digits");
    if year == 0 || NaiveDate::from_ymd_opt(year, month, day).is_none() {
        return Err(ProtocolError("invalid_date"));
    }
    Ok(())
}

pub fn validate_timestamp(value: &str) -> Result<()> {
    let bytes = value.as_bytes();
    if bytes.len() != 24
        || bytes[10] != b'T'
        || bytes[13] != b':'
        || bytes[16] != b':'
        || bytes[19] != b'.'
        || bytes[23] != b'Z'
        || ![0..4, 5..7, 8..10, 11..13, 14..16, 17..19, 20..23]
            .into_iter()
            .all(|range| bytes[range].iter().all(u8::is_ascii_digit))
    {
        return Err(ProtocolError("invalid_timestamp"));
    }
    validate_date(&value[0..10]).map_err(|_| ProtocolError("invalid_timestamp"))?;
    let hour = value[11..13].parse::<u32>().expect("ASCII digits");
    let minute = value[14..16].parse::<u32>().expect("ASCII digits");
    let second = value[17..19].parse::<u32>().expect("ASCII digits");
    if hour > 23 || minute > 59 || second > 59 {
        return Err(ProtocolError("invalid_timestamp"));
    }
    Ok(())
}

pub fn validate_commit_dot(dot: &CommitDot) -> Result<()> {
    validate_canonical_uuid_v4(&dot.writer_id)?;
    parse_writer_seq(&dot.writer_seq)?;
    validate_canonical_uuid_v4(&dot.commit_id)
}

pub fn validate_commit_ref(reference: &CommitRef) -> Result<()> {
    validate_commit_dot(&reference.dot())?;
    if reference.content_hash.len() != 64
        || !reference
            .content_hash
            .as_bytes()
            .iter()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(byte))
    {
        return Err(ProtocolError("invalid_content_hash"));
    }
    Ok(())
}

pub fn compare_commit_dot_v1(a: &CommitDot, b: &CommitDot) -> Ordering {
    a.writer_id
        .as_bytes()
        .cmp(b.writer_id.as_bytes())
        .then_with(|| {
            parse_writer_seq(&a.writer_seq)
                .expect("validated CommitDot")
                .cmp(&parse_writer_seq(&b.writer_seq).expect("validated CommitDot"))
        })
        .then_with(|| a.commit_id.as_bytes().cmp(b.commit_id.as_bytes()))
}

pub fn compare_commit_ref_v1(a: &CommitRef, b: &CommitRef) -> Ordering {
    compare_commit_dot_v1(&a.dot(), &b.dot())
        .then_with(|| a.content_hash.as_bytes().cmp(b.content_hash.as_bytes()))
}

pub fn canonical_entity_key_bytes(key: &EntityKey) -> Result<Vec<u8>> {
    jcs_bytes(key)
}

pub fn compare_entity_key_v1(a: &EntityKey, b: &EntityKey) -> Ordering {
    canonical_entity_key_bytes(a)
        .expect("validated EntityKey")
        .cmp(&canonical_entity_key_bytes(b).expect("validated EntityKey"))
}

fn validate_identity_part(value: &Value) -> Result<()> {
    let text = value.as_str().ok_or(ProtocolError("invalid_entity_key"))?;
    let boundary_whitespace =
        |character: char| matches!(character as u32, 0x09 | 0x0a | 0x0b | 0x0c | 0x0d | 0x20);
    if text.is_empty()
        || text.chars().count() > 512
        || text.contains('\0')
        || text.chars().next().is_some_and(boundary_whitespace)
        || text.chars().next_back().is_some_and(boundary_whitespace)
    {
        return Err(ProtocolError("invalid_entity_key"));
    }
    Ok(())
}

pub fn validate_entity_key(value: &Value) -> Result<()> {
    let parts = value
        .as_array()
        .ok_or(ProtocolError("invalid_entity_key"))?;
    match parts.first().and_then(Value::as_str) {
        Some("record" | "collection") if parts.len() == 2 => validate_identity_part(&parts[1]),
        Some("episode-completion") if parts.len() == 3 => {
            validate_identity_part(&parts[1])?;
            validate_safe_integer(&parts[2], 1, i32::MAX as i64).map(|_| ())
        }
        Some("collection-member") if parts.len() == 3 => {
            validate_identity_part(&parts[1])?;
            validate_identity_part(&parts[2])
        }
        _ => Err(ProtocolError("invalid_entity_key")),
    }
}

pub fn canonical_collection_name_v1(value: &str) -> Result<String> {
    let mut output = String::new();
    let mut pending_space = false;
    for character in value.chars() {
        let code = character as u32;
        let is_s2_whitespace = matches!(code, 0x09 | 0x0a | 0x0b | 0x0c | 0x0d | 0x20);
        if is_s2_whitespace {
            if !output.is_empty() {
                pending_space = true;
            }
            continue;
        }
        if code <= 0x1f || code == 0x7f {
            return Err(ProtocolError("invalid_collection_name"));
        }
        if pending_space {
            output.push(' ');
            pending_space = false;
        }
        output.push(character);
    }
    if output.is_empty() || output.chars().count() > 80 {
        return Err(ProtocolError("invalid_collection_name"));
    }
    Ok(output)
}

pub fn normalize_collection_name_v1(value: &str) -> String {
    value
        .chars()
        .map(|character| {
            if character.is_ascii_uppercase() {
                character.to_ascii_lowercase()
            } else {
                character
            }
        })
        .collect()
}

File: src/features/sync/s2lite/canonical.ts
import type { CommitDot, CommitRef, EntityKey, JsonValue } from './types.ts';

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const UUID_V4_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const UNSIGNED_DECIMAL_RE = /^(0|[1-9][0-9]*)$/;
const SIGNED_DECIMAL_RE = /^-?(0|[1-9][0-9]*)$/;
const SHA256_RE = /^[0-9a-f]{64}$/;
const DATE_RE = /^(\d{4})-(\d{2})-(\d{2})$/;
const TIMESTAMP_RE = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d{3})Z$/;
const S2_WS = new Set([0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x20]);
const U64_MAX = 18_446_744_073_709_551_615n;
const I64_MIN = -9_223_372_036_854_775_808n;
const I64_MAX = 9_223_372_036_854_775_807n;

export class S2ProtocolValidationError extends Error {
  readonly code: string;

  constructor(code: string) {
    super(code);
    this.code = code;
    this.name = 'S2ProtocolValidationError';
  }
}

function invalid(code: string): never {
  throw new S2ProtocolValidationError(code);
}

export function hasLoneSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) return true;
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return true;
    }
  }
  return false;
}

export function validateUnicodeScalarString(value: unknown, code = 'invalid_string'): asserts value is string {
  if (typeof value !== 'string' || value.includes('\0') || hasLoneSurrogate(value)) invalid(code);
}

export function unicodeScalarLength(value: string): number {
  return [...value].length;
}

export function hasBoundaryS2Whitespace(value: string): boolean {
  if (value.length === 0) return false;
  const first = value.codePointAt(0);
  const last = value.codePointAt(value.length - 1);
  return (first !== undefined && S2_WS.has(first)) || (last !== undefined && S2_WS.has(last));
}

function compareUtf16(a: string, b: string): number {
  const length = Math.min(a.length, b.length);
  for (let index = 0; index < length; index += 1) {
    const difference = a.charCodeAt(index) - b.charCodeAt(index);
    if (difference !== 0) return Math.sign(difference);
  }
  return Math.sign(a.length - b.length);
}

export function canonicalizeJcs(value: JsonValue): string {
  if (value === null) return 'null';
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (typeof value === 'string') {
    validateUnicodeScalarString(value, 'jcs_invalid_string');
    return JSON.stringify(value);
  }
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) invalid('jcs_non_finite_number');
    return JSON.stringify(Object.is(value, -0) ? 0 : value);
  }
  if (Array.isArray(value)) {
    const items: string[] = [];
    for (let index = 0; index < value.length; index += 1) {
      if (!Object.prototype.hasOwnProperty.call(value, index)) invalid('jcs_sparse_array');
      items.push(canonicalizeJcs(value[index]!));
    }
    return `[${items.join(',')}]`;
  }
  if (typeof value !== 'object') return invalid('jcs_unsupported_value');

  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) invalid('jcs_non_plain_object');
  if (Reflect.ownKeys(value).some(key => typeof key === 'symbol')) invalid('jcs_unsupported_value');
  const keys = Object.keys(value).sort(compareUtf16);
  return `{${keys.map(key => `${canonicalizeJcs(key)}:${canonicalizeJcs(value[key]!)}`).join(',')}}`;
}

export function canonicalJcsBytes(value: JsonValue): Uint8Array {
  return new TextEncoder().encode(canonicalizeJcs(value));
}

export async function sha256Hex(bytes: Uint8Array): Promise<string> {
  const exactBytes = Uint8Array.from(bytes);
  const digest = await crypto.subtle.digest('SHA-256', exactBytes.buffer as ArrayBuffer);
  return Array.from(new Uint8Array(digest), byte => byte.toString(16).padStart(2, '0')).join('');
}

export async function sha256Jcs(value: JsonValue): Promise<string> {
  return sha256Hex(canonicalJcsBytes(value));
}

export function validateCanonicalUuid(value: unknown): asserts value is string {
  if (typeof value !== 'string' || !UUID_RE.test(value)) invalid('invalid_canonical_uuid');
}

export function validateCanonicalUuidV4(value: unknown): asserts value is string {
  if (typeof value !== 'string' || !UUID_V4_RE.test(value)) invalid('invalid_canonical_uuid_v4');
}

export function validateContentHash(value: unknown): asserts value is string {
  if (typeof value !== 'string' || !SHA256_RE.test(value)) invalid('invalid_content_hash');
}

export function parseWriterSeq(value: unknown): bigint {
  if (typeof value !== 'string' || !UNSIGNED_DECIMAL_RE.test(value)) return invalid('invalid_writer_seq');
  const parsed = BigInt(value);
  if (parsed < 1n || parsed > U64_MAX) return invalid('invalid_writer_seq');
  return parsed;
}

export function parseInt64DecimalString(
  value: unknown,
  minimum = I64_MIN,
  maximum = I64_MAX,
): bigint {
  if (typeof value !== 'string' || !SIGNED_DECIMAL_RE.test(value) || value === '-0') {
    return invalid('invalid_int64_decimal_string');
  }
  const parsed = BigInt(value);
  if (parsed < minimum || parsed > maximum) return invalid('int64_out_of_range');
  return parsed;
}

export function validateSafeInteger(value: unknown, minimum: number, maximum: number): asserts value is number {
  if (!Number.isSafeInteger(value) || (value as number) < minimum || (value as number) > maximum) {
    invalid('invalid_safe_integer');
  }
}

export function extractSafeIntegerV1(value: unknown, minimum: number, maximum: number): number {
  validateSafeInteger(value, minimum, maximum);
  return value;
}

export function validateFloat64(value: unknown, minimum: number, maximum: number): asserts value is number {
  if (typeof value !== 'number' || !Number.isFinite(value) || value < minimum || value > maximum) {
    invalid('invalid_float64');
  }
}

function isLeapYear(year: number): boolean {
  return year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
}

function validDateParts(year: number, month: number, day: number): boolean {
  if (year < 1 || year > 9999 || month < 1 || month > 12) return false;
  const days = [31, isLeapYear(year) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  return day >= 1 && day <= days[month - 1]!;
}

export function validateCanonicalDate(value: unknown): asserts value is string {
  if (typeof value !== 'string') invalid('invalid_date');
  const match = DATE_RE.exec(value);
  if (!match || !validDateParts(Number(match[1]), Number(match[2]), Number(match[3]))) invalid('invalid_date');
}

export function validateCanonicalTimestamp(value: unknown): asserts value is string {
  if (typeof value !== 'string') invalid('invalid_timestamp');
  const match = TIMESTAMP_RE.exec(value);
  if (!match || !validDateParts(Number(match[1]), Number(match[2]), Number(match[3]))) {
    invalid('invalid_timestamp');
  }
  const hour = Number(match[4]);
  const minute = Number(match[5]);
  const second = Number(match[6]);
  if (hour > 23 || minute > 59 || second > 59) invalid('invalid_timestamp');
}

export function normalizeCollectionNameV1(value: string): string {
  let result = '';
  for (const character of value) {
    const codePoint = character.codePointAt(0)!;
    result += codePoint >= 0x41 && codePoint <= 0x5a
      ? String.fromCodePoint(codePoint + 0x20)
      : character;
  }
  return result;
}

export function canonicalCollectionNameV1(value: unknown): string {
  validateUnicodeScalarString(value, 'invalid_collection_name');
  let output = '';
  let pendingSpace = false;
  for (const character of value) {
    const codePoint = character.codePointAt(0)!;
    if (S2_WS.has(codePoint)) {
      if (output.length > 0) pendingSpace = true;
      continue;
    }
    if ((codePoint >= 0 && codePoint <= 0x1f) || codePoint === 0x7f) {
      invalid('invalid_collection_name');
    }
    if (pendingSpace) {
      output += ' ';
      pendingSpace = false;
    }
    output += character;
  }
  if (unicodeScalarLength(output) < 1 || unicodeScalarLength(output) > 80) invalid('invalid_collection_name');
  return output;
}

function compareAscii(a: string, b: string): number {
  const length = Math.min(a.length, b.length);
  for (let index = 0; index < length; index += 1) {
    const difference = a.charCodeAt(index) - b.charCodeAt(index);
    if (difference !== 0) return Math.sign(difference);
  }
  return Math.sign(a.length - b.length);
}

export function compareCommitDotV1(a: CommitDot, b: CommitDot): number {
  let comparison = compareAscii(a.writerId, b.writerId);
  if (comparison !== 0) return comparison;
  const aSeq = parseWriterSeq(a.writerSeq);
  const bSeq = parseWriterSeq(b.writerSeq);
  comparison = aSeq < bSeq ? -1 : aSeq > bSeq ? 1 : 0;
  return comparison || compareAscii(a.commitId, b.commitId);
}

export function compareCommitRefV1(a: CommitRef, b: CommitRef): number {
  const dotComparison = compareCommitDotV1(a, b);
  return dotComparison || compareAscii(a.contentHash, b.contentHash);
}

export function compareUnsignedBytes(a: Uint8Array, b: Uint8Array): number {
  const length = Math.min(a.length, b.length);
  for (let index = 0; index < length; index += 1) {
    const difference = a[index]! - b[index]!;
    if (difference !== 0) return Math.sign(difference);
  }
  return Math.sign(a.length - b.length);
}

export function canonicalEntityKeyBytes(key: EntityKey): Uint8Array {
  return canonicalJcsBytes(key as unknown as JsonValue);
}

export function compareEntityKeyV1(a: EntityKey, b: EntityKey): number {
  return compareUnsignedBytes(canonicalEntityKeyBytes(a), canonicalEntityKeyBytes(b));
}

function validateIdentityPart(value: unknown): void {
  validateUnicodeScalarString(value, 'invalid_entity_key');
  if (value.length === 0 || unicodeScalarLength(value) > 512 || hasBoundaryS2Whitespace(value)) {
    invalid('invalid_entity_key');
  }
}

export function validateEntityKey(value: unknown): asserts value is EntityKey {
  if (!Array.isArray(value) || typeof value[0] !== 'string') invalid('invalid_entity_key');
  if (value[0] === 'record' || value[0] === 'collection') {
    if (value.length !== 2) invalid('invalid_entity_key');
    validateIdentityPart(value[1]);
  } else if (value[0] === 'episode-completion') {
    if (value.length !== 3) invalid('invalid_entity_key');
    validateIdentityPart(value[1]);
    validateSafeInteger(value[2], 1, 2_147_483_647);
  } else if (value[0] === 'collection-member') {
    if (value.length !== 3) invalid('invalid_entity_key');
    validateIdentityPart(value[1]);
    validateIdentityPart(value[2]);
  } else {
    invalid('invalid_entity_key');
  }
}

function hasExactObjectFields(value: unknown, expected: readonly string[]): value is Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) return false;
  const actual = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return actual.length === sortedExpected.length
    && actual.every((field, index) => field === sortedExpected[index]);
}

export function validateCommitDot(dot: CommitDot): void {
  if (!hasExactObjectFields(dot, ['writerId', 'writerSeq', 'commitId'])) invalid('invalid_commit_dot');
  validateCanonicalUuidV4(dot.writerId);
  parseWriterSeq(dot.writerSeq);
  validateCanonicalUuidV4(dot.commitId);
}

export function validateCommitRef(ref: CommitRef): void {
  if (!hasExactObjectFields(ref, ['writerId', 'writerSeq', 'commitId', 'contentHash'])) {
    invalid('invalid_commit_ref');
  }
  validateCanonicalUuidV4(ref.writerId);
  parseWriterSeq(ref.writerSeq);
  validateCanonicalUuidV4(ref.commitId);
  validateContentHash(ref.contentHash);
}

export const INT64_MIN = I64_MIN;
export const INT64_MAX = I64_MAX;
export const UINT64_MAX = U64_MAX;
